File: pkg/api/transformer/place_card.go
package transformer

import (
    "encoding/json"
    "strconv"

    "github.com/alirezadp10/hokm/pkg/service"
)

type PlaceCardTransformerData struct {
    Players           []string
    UIndex            int
    Points            string
    Cards             map[int][]string
    CenterCards       string
    Turn              string
    King              string
    Trump             string
    LastMoveTimestamp string
    WasTheKingChanged string
    CardsWinner       string
    RoundWinner       string
    GameWinner        string
    LeadSuit          string
}

func PlaceCardTransformer(playersService *service.PlayersService, cardsService *service.CardsService, data PlaceCardTransformerData) map[string]interface{} {
    cards, _ := json.Marshal(data.Cards)
    result := map[string]interface{}{
        "players":           playersService.GetPlayersWithDirections(data.Players, data.UIndex),
        "points":            cardsService.GetPoints(data.Points, data.UIndex),
        "centerCards":       playersService.GetPlayersCenterCards(data.CenterCards, data.UIndex),
        "turn":              playersService.GetTurn(data.Turn, data.UIndex),
        "king":              playersService.GetKing(data.King, data.UIndex),
        "timeRemained":      playersService.GetTimeRemained(data.LastMoveTimestamp),
        "playerCards":       cardsService.GetPlayerCards(string(cards), data.UIndex),
        "WasTheKingChanged": data.WasTheKingChanged,
        "trump":             data.Trump,
        "leadSuit":          data.LeadSuit,
        "whoHasWonTheCards": "",
        "whoHasWonTheRound": "",
        "whoHasWonTheGame":  "",
    }

    if data.CardsWinner != "" {
        cardsWinner, _ := strconv.Atoi(data.CardsWinner)
        result["whoHasWonTheCards"] = playersService.GetDirection(cardsWinner, data.UIndex)
    }

    if data.RoundWinner != "" {
        roundWinner, _ := strconv.Atoi(data.RoundWinner)
        result["whoHasWonTheRound"] = playersService.GetDirection(roundWinner, data.UIndex)
    }

    if data.GameWinner != "" {
        gameWinner, _ := strconv.Atoi(data.GameWinner)
        result["whoHasWonTheGame"] = playersService.GetDirection(gameWinner, data.UIndex)
    }

    return result
}

File: pkg/repository/redis/lua/place-card.lua
redis.call('HSET', "game:" .. KEYS[1], 'center_cards', ARGV[1])
redis.call('HSET', "game:" .. KEYS[1], 'lead_suit', ARGV[2])
redis.call('HSET', "game:" .. KEYS[1], 'who_has_won_the_cards', ARGV[3])
redis.call('HSET', "game:" .. KEYS[1], 'points', ARGV[4])
redis.call('HSET', "game:" .. KEYS[1], 'turn', ARGV[5])
redis.call('HSET', "game:" .. KEYS[1], 'king', ARGV[6])
redis.call('HSET', "game:" .. KEYS[1], 'was_the_king_changed', ARGV[7])
redis.call('HSET', "game:" .. KEYS[1], 'last_move_timestamp', ARGV[14])
redis.call('HSET', "game:" .. KEYS[1], 'trump', ARGV[15])
redis.call('HSET', "game:" .. KEYS[1], 'cards', cjson.encode({
    [0] = cjson.decode(ARGV[8]),
    [1] = cjson.decode(ARGV[9]),
    [2] = cjson.decode(ARGV[10]),
    [3] = cjson.decode(ARGV[11])
}))
redis.call('PUBLISH', "placing_card", KEYS[1] .. "|" .. ARGV[12] .. "|" .. ARGV[13])
return ""